### PR TITLE
Make three silent failures in the Raiden training path visible

### PR DESCRIPTION
**Pull request by A9isha**
# Description

Each change targets a failure that produced no usable signal at the point of cause: weights that never transfer, metadata that fails in another process, and a performance cliff recorded only in a docstring.

unscan_layers had no tests, and it is the piece that decides whether trainer and sampler tensor names agree. Both sides name tensors with jax.tree_util.keystr, and nothing cross-checks the two sets: raiden_handler._validate_metadata only validates one manifest's internal consistency (mesh rank, duplicate variable/layer keys, sharding specs). A naming error therefore surfaces as weights that silently never transfer. Two cases pin non-obvious invariants: unscan_layers returns a plain nested dict while the sampler binds an nnx.State, and keystr renders those identically only because the transform rewraps leaves in nnx.Param -- dropping that rewrap would rename every tensor ("['k']" vs "['k'].value"); and an already-unscanned state must raise, since without that guard it would return unchanged and bind under scanned names.

prepare_weight_sync returned empty metadata on two paths: a missing raiden_synchronizer (warning-level) and an unrecognised staging_transport (no log at all). Neither is silent end to end -- WeightSyncCoordinator rejects an empty side -- but the failure lands far from the cause, surfacing in another process as "metadata collection returned an empty side", a count that never names the missing module or the bad transport. The import case is the common one rather than a corner: raiden_synchronizer ships only on tunix's Raiden branch, so any released tunix takes it. Both now raise where the cause is known, with the ImportError chained so the traceback keeps the module name. Because staging_transport defaults to "raiden", this also reaches callers that never asked for it, so the engine e2e test now probes for the synchronizer the way the engine does -- exercising the real staging path where Raiden exists and the documented failure where it does not.

_batch_data_shardings falls back to replicating the batch dimension when it does not divide the batch axis's mesh size. That is correct -- every device along the axis computes the whole micro-batch -- but it costs N times the work a sharded one would do there. An invisible performance cliff is harder to notice than a wrong number, because XLA's caching can make it look like nothing worse than a slow run; the file already warns once per instance when a signature half cannot be compared, and this extends that treatment. Warned once per instance rather than per leaf, since the check runs under a tree_map over every loss input and they normally share a batch dim. A sequence-packed micro-batch is always size 1 and has no alternative, so the message says the fallback may well be deliberate.

Verification. The unscan suite has teeth: renaming the emitted key from layers_{i} to layer_{i} fails 5 of its 11 tests, including the name-equality one. Marked post_training and left in tests/unit, which is already in cpu-post-training-unit's path list, so the marker alone routes it -- tests/ and tests/integration are not in that list, which is how the engine tests once ended up collected by no job at all; collection confirms 11 tests in cpu-post-training-unit and 0 in cpu-unit. The staging and sharding tests fail without their respective changes. The sharding tests stub both the data spec and the axis size: a single-device test mesh returns None in the batch position, making the branch unreachable as configured, and an earlier draft asserted `spec[0] is None` and passed without running the code under test at all.

# Tests

Ran on a TPU-VM v5p-8 and also on v5p-128 cluster

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by igorts-git — APPROVED** [review]

File: tests/post_training/unit/maxtext_engine_test.py [hunk 1]
```diff
@@ -28,6 +29,7 @@
 from maxtext.training_engine import abstract_engine
 from maxtext.training_engine import maxtext_engine
 from maxtext.utils import maxtext_utils
+from maxtext.utils import sharding
```
[the remark this answers is not in the document]
Comment (A9isha) [reply tests/post_training/unit/maxtext_engine_test.py]: done

File: tests/unit/raiden_unscan_test.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+
+_NUM_LAYERS = 3
+_IN, _OUT, _VOCAB = 4, 8, 10
+
+
+def _unwrap(leaf):
+  """Reads a leaf's array whether or not it is wrapped in an `nnx.Param`."""
+  if isinstance(leaf, nnx.Variable):
+    return leaf[...]
```
[the remark this answers is not in the document]
Comment (A9isha) [reply tests/unit/raiden_unscan_test.py:49]: Keeping `leaf[...]` here. Flax has deprecated `.value` in the opposite
direction from what this suggests. On flax 0.12.8, `Variable.value` emits:

  DeprecationWarning: '.value' access is now deprecated.
  For Variable[Array] instances use:  variable[...]
  For other Variable types use:       variable.get_value()

and `HijaxVariable.value` (flax/nnx/variablelib.py:696) doesn't warn but raises
NotImplementedError, also pointing at `variable[...]`. So indexing is the
current idiom, and it's the more robust of the two for a helper that has to
accept any Variable subtype — `.value` would add a DeprecationWarning to every
call today and fail outright on HijaxVariable.

**Review by YixuanWang-99 — APPROVED** [review]
